GPTutils.py

```diff
@@ -0,0 +1,262 @@
+    """
+    return np.dot(a, b.T) / (norm(a) * norm(b, axis=1))
+
+
+class FilesIndex:
+    def __init__(self, files: dict[str, str]) -> None:
+        """
+        Initializes a new instance of the FilesIndex class.
+
+        Args:
+            files (dict[str, str]): A dictionary containing the files to be indexed.
+            {Filename: FileContent}
+        """
+        self.files = files
+        self.index: Dict[str, List[int]] = {}
+        self.idx_to_filename = {}
+        self.embeddings_model = OpenAIEmbeddings()
+        self.docs: List[str] = []
+        self.embeddings: np.ndarray = None
+        self.index_files()
+
+    def index_files(self) -> None:
+        """
+        Indexes the files in the `files` dictionary by creating a mapping between each file name and a list of document
+        indices in the `docs` list. Also computes embeddings for each document and stores them in the `embeddings` array.
+        """
+        embeddings = []
+        last_index = 0
+        for filename, content in self.files.items():
+            extension = filename.split(".")[-1]
+            docs = to_docs(content, extension)
+            idxs = list(range(last_index, last_index + len(docs)))
+            last_index += len(docs)
+            self.index[filename] = idxs
+            self.docs += docs
+            embeddings += self.embeddings_model.embed_documents(docs)
+        self.embeddings = np.array(embeddings)
+        for filename, idxs in self.index.items():
+            for idx in idxs:
+                self.idx_to_filename[idx] = filename
+
+    def search_docs(
+        self, queries: Union[str, List[str]], top_k: int = 5, sorted_by="score"
+    ) -> Tuple[List[int], List[str]]:
+        """
+        Searches for the top k documents that are most similar to the given queries.
+
+
+        Args:
+            query (str): The query to search for.
+            top_k (int, optional): The number of top documents to return. Defaults to 5.
+            sorted_by (str, optional)
+        Returns:
+            List[str]: A list of the top k documents that are most similar to the given query.
+        """
+        if isinstance(queries, list):
+            query_embeddings = []
+            for query in queries:
+                query_embeddings.append(
+                    np.array(self.embeddings_model.embed_query(query))
+                )
+
+            scores = cosim_matrix(query_embeddings[0], self.embeddings)
+            for query_embedding in query_embeddings[1:]:
+                scores += cosim_matrix(query_embedding, self.embeddings)
+
+        elif isinstance(queries, str):
+            query = queries  # signaling that queries is a string
+            query_embedding = np.array(self.embeddings_model.embed_query(query))
+            scores = cosim_matrix(query_embedding, self.embeddings)
+
+        top_k_idxs = np.argsort(scores)[::-1][:top_k]
+
+        if sorted_by == "file":
+            top_k_idxs = sorted(top_k_idxs)
+
```

Comment:
Incorporate exception catching for OpenAI client interactions to avoid unexpected crashes and to handle API limitations or downtime gracefully, for example:
```suggestion
    def get_response(self, max_tokens: int = 1000) -> str:
        try:
            response = self.client.chat.completions.create(
                model=self.model, messages=self.conversation, max_tokens=max_tokens
            )
            return response.choices[0].message.content
        except openai.OpenAIError as e:
            logging.error(f"OpenAI API error: {e}")
            return ""
```